bundles/org.openhab.binding.flume/src/main/java/org/openhab/binding/flume/internal/api/FlumeApi.java

```diff
@@ -0,0 +1,411 @@
+        boolean success = jsonResponse.get("success").getAsBoolean();
+
+        if (!success) {
+            String message = jsonResponse.get("message").getAsString();
+            throw new FlumeApiException("@text/api.query-fail [\"" + message + "\"]",
+                    jsonResponse.get("code").getAsInt(), false);
+        }
+
+        final FlumeApiToken[] data = gson.fromJson(jsonResponse.get("data").getAsJsonArray(), FlumeApiToken[].class);
+
+        if (data == null) {
+            throw new FlumeApiException("@text/api.response-invalid", jsonResponse.get("code").getAsInt(), true);
+        }
+
+        processToken(data[0]);
+    }
+
+    private void refreshToken()
+            throws IOException, InterruptedException, TimeoutException, ExecutionException, FlumeApiException {
+        FlumeApiRefreshToken token = new FlumeApiRefreshToken();
+
+        token.clientId = clientId;
+        token.clientSecret = clientSecret;
+        token.refeshToken = refreshToken;
+
+        String url = APIURL_BASE + APIURL_TOKEN;
+        Request request = httpClient.newRequest(url).method(HttpMethod.POST)
+                .header(HttpHeader.CONTENT_TYPE, "application/json").header("Authorization", "Bearer " + accessToken)
+                .content(new StringContentProvider(gson.toJson(token)), "application/json");
+
+        logger.trace("POST: {}", url);
+        ContentResponse response = request.send();
+        JsonObject jsonResponse = validateResponse(response);
+
+        final FlumeApiToken[] data = gson.fromJson(jsonResponse.get("data").getAsJsonArray(), FlumeApiToken[].class);
+
+        if (data == null) {
+            throw new FlumeApiException("@text/api.response-invalid", jsonResponse.get("code").getAsInt(), true);
+        }
+
+        processToken(data[0]);
+    }
+
+    private void processToken(FlumeApiToken token) throws FlumeApiException {
+        accessToken = token.accessToken;
+
+        // access_token contains 3 parts: header, payload, signature - decode the payload portion
+        String accessTokenPayload[] = accessToken.split("\\.");
+        byte decoded[] = Base64.getDecoder().decode(accessTokenPayload[1]);
+
+        String jsonPayload = new String(decoded);
+
+        FlumeApiTokenPayload payload = gson.fromJson(jsonPayload, FlumeApiTokenPayload.class);
+
+        if (payload == null) {
+            throw new FlumeApiException("@text/api.response-invalid", 0, true);
+        }
+
+        userId = payload.userId;
+
+        refreshToken = token.refreshToken;
+        tokenExpiresAt = LocalDateTime.now().plusSeconds(token.expiresIn);
+
+        logger.debug("Token expires in: {}", tokenExpiresAt);
+    }
+
+    public void verifyToken()
+            throws FlumeApiException, IOException, InterruptedException, TimeoutException, ExecutionException {
+        // refresh token 4 hours before it expires
+        if (tokenExpiresAt.isBefore(LocalDateTime.now().minusHours(4))) {
```

Comment:
I'm used to renew when the token is at 2/3 of its lifetime. Might be perfectly sane, but hardcodeing 4 hours might also cause isseus in thew futer?
Or wdyt?